File: clj/test/cantrip/examples_test.clj
(ns cantrip.examples-test
  "Structural tests for grimoire teaching examples.

   These tests verify that each example demonstrates its pattern correctly,
   regardless of LLM output. They test structure, not content.

   Cross-cutting requirements:
     - (example-NN {:mode :scripted}) uses FakeLLM, works without env vars
     - (example-NN) or (example-NN {:mode :real}) with no env vars MUST throw
     - Every result map has :pattern key matching its number

   These tests MAY fail against current examples -- that is the point.
   They establish what 'correct' looks like per the spec."
  (:require [cantrip.examples :as examples]
            [cantrip.gates :as gates]
            [cantrip.runtime :as runtime]
            [clojure.test :refer [deftest is testing]]))

;; ── Cross-cutting: scripted mode always works ────────────────────────────────

(deftest scripted-mode-01 (is (= 1 (:pattern (examples/example-01-llm-query {:mode :scripted})))))
(deftest scripted-mode-02 (is (= 2 (:pattern (examples/example-02-gate)))))
(deftest scripted-mode-03 (is (= 3 (:pattern (examples/example-03-circle)))))
(deftest scripted-mode-04 (is (= 4 (:pattern (examples/example-04-cantrip {:mode :scripted})))))
(deftest scripted-mode-05 (is (= 5 (:pattern (examples/example-05-wards {:mode :scripted})))))
(deftest scripted-mode-06 (is (= 6 (:pattern (examples/example-06-medium {:mode :scripted})))))
(deftest scripted-mode-07 (is (= 7 (:pattern (examples/example-07-full-agent {:mode :scripted})))))
(deftest scripted-mode-08 (is (= 8 (:pattern (examples/example-08-folding {:mode :scripted})))))
(deftest scripted-mode-09 (is (= 9 (:pattern (examples/example-09-composition {:mode :scripted})))))
(deftest scripted-mode-10 (is (= 10 (:pattern (examples/example-10-loom {:mode :scripted})))))
(deftest scripted-mode-11 (is (= 11 (:pattern (examples/example-11-persistent-entity {:mode :scripted})))))
(deftest scripted-mode-12 (is (= 12 (:pattern (examples/example-12-familiar {:mode :scripted})))))
(deftest scripted-mode-13 (is (= 13 (:pattern (examples/example-13-acp {:mode :scripted})))))

;; ── Cross-cutting: no silent fallback ────────────────────────────────────────
;; Examples 02 and 03 don't need LLM, so they're excluded.

(deftest no-fallback-01 (is (thrown? Exception (examples/example-01-llm-query {:mode :real}))))
(deftest no-fallback-04 (is (thrown? Exception (examples/example-04-cantrip {:mode :real}))))
(deftest no-fallback-05 (is (thrown? Exception (examples/example-05-wards {:mode :real}))))
(deftest no-fallback-06 (is (thrown? Exception (examples/example-06-medium {:mode :real}))))
(deftest no-fallback-07 (is (thrown? Exception (examples/example-07-full-agent {:mode :real}))))
(deftest no-fallback-08 (is (thrown? Exception (examples/example-08-folding {:mode :real}))))
(deftest no-fallback-09 (is (thrown? Exception (examples/example-09-composition {:mode :real}))))
(deftest no-fallback-10 (is (thrown? Exception (examples/example-10-loom {:mode :real}))))
(deftest no-fallback-11 (is (thrown? Exception (examples/example-11-persistent-entity {:mode :real}))))
(deftest no-fallback-12 (is (thrown? Exception (examples/example-12-familiar {:mode :real}))))

;; ── Per-example structural tests (scripted mode) ────────────────────────────

(deftest example-01-llm-query-test
  (let [result (examples/example-01-llm-query {:mode :scripted})]
    (is (= 1 (:pattern result)))
    ;; Stateless: one query, one response, no loop
    (is (= 1 (count (get-in result [:query :messages])))
        "must send exactly one message")
    (is (string? (get-in result [:response :content]))
        "response must contain a content string")))

(deftest example-02-gate-test
  (let [result (examples/example-02-gate)]
    (is (= 2 (:pattern result)))
    ;; Tools list is non-empty
    (is (seq (:tools result))
        "gate-tools must return tools")
    ;; Echo gate works
    (is (= false (get-in result [:echo-exec :observation 0 :is-error]))
        "echo gate call must not be an error")
    ;; Done gate terminates
    (is (true? (:terminated? (:done-exec result)))
        "done gate must terminate the loop")
    ;; Malformed done (empty args) must be error, NOT terminate
    (is (true? (get-in result [:malformed-done :observation 0 :is-error]))
        "malformed done (empty args) must be an error")
    (is (false? (:terminated? (:malformed-done result)))
        "malformed done must NOT terminate")))

(deftest example-03-circle-test
  (let [result (examples/example-03-circle)]
    (is (= 3 (:pattern result)))
    ;; Valid cantrip exists
    (is (map? (:valid result)))
    ;; Missing done gate produces CIRCLE-1 error
    (is (= "CIRCLE-1" (:rule (:missing-done result)))
        "missing done must cite CIRCLE-1")
    ;; Missing wards produces CIRCLE-2 error
    (is (= "CIRCLE-2" (:rule (:missing-wards result)))
        "missing wards must cite CIRCLE-2")))

(deftest example-04-cantrip-test
  (let [result (examples/example-04-cantrip {:mode :scripted})]
    (is (= 4 (:pattern result)))
    ;; Both runs terminated
    (is (= :terminated (:status (:first-run result)))
        "first cast must terminate")
    (is (= :terminated (:status (:second-run result)))
        "second cast must terminate")
    ;; Each run has turns
    (is (pos? (count (:turns (:first-run result))))
        "first run must have turns")
    ;; Independent entity IDs (CANTRIP-2)
    (is (true? (:independent-entity-ids result))
        "two casts must produce independent entity IDs")))

(deftest example-05-wards-test
  (let [result (examples/example-05-wards {:mode :scripted})]
    (is (= 5 (:pattern result)))
    ;; Ward composition: min wins for numeric
    (is (= 10 (get-in result [:composed :max-turns]))
        "composed max-turns must be 10 (min wins)")
    ;; Ward composition: OR wins for boolean
    (is (true? (get-in result [:composed :require-done-tool]))
        "require-done-tool must be true (OR wins)")
    ;; Run should be truncated (entity echoes but hits ward before done)
    (is (= :truncated (:status (:run result)))
        "run must be truncated (ward cuts off before done)")))

(deftest example-06-medium-test
  (let [result (examples/example-06-medium {:mode :scripted})]
    (is (= 6 (:pattern result)))
    ;; Two different mediums
    (is (= :conversation (get-in result [:conversation :view :medium]))
        "conversation view must have :conversation medium")
    (is (= :code (get-in result [:code :view :medium]))
        "code view must have :code medium")
    ;; Both runs terminate
    (is (= :terminated (get-in result [:conversation :run :status]))
        "conversation run must terminate")
    (is (= :terminated (get-in result [:code :run :status]))
        "code run must terminate")))

(deftest example-07-full-agent-test
  (let [result (examples/example-07-full-agent {:mode :scripted})]
    (is (= 7 (:pattern result)))
    ;; Terminated
    (is (= :terminated (:status (:run result)))
        "agent must terminate")
    ;; At least 2 turns for error + recovery
    (is (>= (count (:turns (:run result))) 2)
        "need >= 2 turns")
    ;; Error steering: some observation is an error
    (is (some :is-error (:observations result))
        "at least one observation must be an error")
    ;; Recovery: done gate called
    (is (some #(= "done" %) (:gate-seq result))
        "gate sequence must include done")
    ;; DEEP CHECK: error-then-recovery ordering
    (let [obs (vec (:observations result))]
      (when (>= (count obs) 2)
        (is (true? (:is-error (first obs)))
            "first observation must be an error")
        (is (false? (:is-error (last obs)))
            "last observation must NOT be an error (recovery)")))))

(deftest example-08-folding-test
  (let [result (examples/example-08-folding {:mode :scripted})]
    (is (= 8 (:pattern result)))
    ;; 3 invocations (one per send)
    (is (= 3 (count (:invocations result)))
        "must have 3 LLM invocations")
    ;; State has 3 turns
    (is (= 3 (:turn-count (:state result)))
        "state must have turn-count 3")
    ;; Folding markers present and contain "Folded" text
    (is (seq (:folding-markers result))
        "folding markers must be non-empty")
    (is (every? #(re-find #"(?i)folded" (str %)) (:folding-markers result))
        "each folding marker must contain 'Folded' text")
    ;; Identity (system prompt) preserved through folding
    (is (string? (get-in result [:state :loom :identity :system-prompt]))
        "system prompt must be preserved in loom")))

(deftest example-09-composition-test
  (let [result (examples/example-09-composition {:mode :scripted})]
    (is (= 9 (:pattern result)))
    ;; Single child terminated
    (is (= :terminated (:status (:single result)))
        "single child must terminate")
    ;; Batch has 2 results, all terminated
    (is (= 2 (count (:batch result)))
        "batch must have 2 results")
    (is (every? #(= :terminated (:status %)) (:batch result))
        "all batch results must terminate")
    ;; Parent state has delegation turns (>= 3: intent + call + batch)
    (is (>= (count (get-in result [:parent-state :loom :turns])) 3)
        "parent loom must have >= 3 turns")))

(deftest example-10-loom-test
  (let [result (examples/example-10-loom {:mode :scripted})]
    (is (= 10 (:pattern result)))
    ;; Terminated
    (is (= :terminated (:status result))
        "run must terminate")
    ;; Turn counts consistent
    (is (pos? (:turn-count result))
        "must have positive turn count")
    (is (= (:turn-count result) (:loom-turn-count result))
        "turn count must match loom turn count")
    ;; Token usage tracked
    (is (map? (:token-usage result))
        "token usage must be a map")
    ;; Gates called
    (is (seq (:gates-called result))
        "gates-called must be non-empty")
    (is (some #(= "echo" %) (:gates-called result))
        "echo must be in gates-called")
    (is (some #(= "done" %) (:gates-called result))
        "done must be in gates-called")))

(deftest example-11-persistent-entity-test
  (let [result (examples/example-11-persistent-entity {:mode :scripted})]
    (is (= 11 (:pattern result)))
    ;; Both sends terminated
    (is (= :terminated (:status (:first-send result)))
        "first send must terminate")
    (is (= :terminated (:status (:second-send result)))
        "second send must terminate")
    ;; State accumulates: 2 turns total
    (is (= 2 (:turn-count (:state result)))
        "entity must have 2 accumulated turns")
    ;; Loom has 2 turns
    (is (= 2 (count (get-in result [:state :loom :turns])))
        "loom must have 2 turns")))

(deftest example-12-familiar-test
  (let [result (examples/example-12-familiar {:mode :scripted})]
    (is (= 12 (:pattern result)))
    ;; Both sends terminated
    (is (= :terminated (:status (:first-send result)))
        "first send must terminate")
    (is (= :terminated (:status (:second-send result)))
        "second send must terminate")
    ;; State accumulates
    (is (>= (:turn-count (:state result)) 2)
        "state must have >= 2 turns")
    ;; Loom turns exist
    (is (seq (get-in result [:state :loom :turns]))
        "loom turns must exist")
    ;; First send result must contain child delegation evidence
    (let [first-result (get-in result [:first-send :result])]
      (is (string? first-result) "first send result must be a string")
      (is (re-find #"child" (str first-result))
          "first send result should mention child results (evidence of delegation)"))))

(deftest example-13-acp-test
  (let [result (examples/example-13-acp {:mode :scripted})]
    (is (= 13 (:pattern result)))
    (is (string? (:session-id result))
        "session-id must be a string")
    (is (= "2.0" (get-in result [:response :jsonrpc]))
        "response must have jsonrpc 2.0")
    (is (seq (get-in result [:response :result :output]))
        "response must have output")))

;; ── Framework-level structural checks ────────────────────────────────────────

(deftest done-gate-has-parameter-schema
  (testing "done gate must have answer parameter in schema"
    (let [tools (gates/gate-tools [:done :echo])
          done-tool (first (filter #(= "done" (:name %)) tools))]
      (is (some? done-tool)
          "done must appear in gate-tools output")
      (is (map? (:parameters done-tool))
          "done gate must have :parameters map")
      (when (map? (:parameters done-tool))
        (let [props (or (get-in done-tool [:parameters :properties])
                        (get-in done-tool [:parameters "properties"]))]
          (is (some? props)
              "done parameters must have properties")
          (when props
            (is (or (contains? props :answer) (contains? props "answer"))
                "done properties must include 'answer'")))))))

(deftest child-identity-not-parent-delegation
  (testing "child entity must NOT inherit parent's delegation-specific identity"
    ;; When derive-child-cantrip produces a child, it should get a generic
    ;; identity unless one is explicitly provided, not the parent's prompt
    ;; about delegation gates it doesn't have.
    (let [parent (runtime/summon
                  {:llm {:provider :fake
                         :responses [{:tool-calls [{:id "p1" :gate :done :args {:answer "parent"}}]}]}
                   :identity {:system-prompt "I am the parent. Delegate tasks using call-agent."}
                   :circle {:medium :conversation
                            :gates [:done]
                            :wards [{:max-turns 2} {:max-depth 2}]}})
          ;; When providing an explicit child cantrip, it keeps its identity
          child-cantrip {:llm {:provider :fake
                               :responses [{:tool-calls [{:id "c1" :gate :done :args {:answer "child"}}]}]}
                         :identity {:system-prompt "I am a child worker."}
                         :circle {:medium :conversation
                                  :gates [:done]
                                  :wards [{:max-turns 2}]}}
          result (runtime/call-agent parent {:cantrip child-cantrip :intent "child task"})]
      (is (= :terminated (:status result))
          "child must terminate"))))

(deftest pattern-notes-coverage-test
  (is (= (set (map #(format "%02d" %) (range 1 14)))
         (set (keys examples/pattern-notes)))))

File: ex/test/examples_test.exs
defmodule CantripExamplesTest do
  @moduledoc """
  Structural tests for grimoire teaching examples.

  These tests verify that each example demonstrates its pattern correctly,
  regardless of LLM output. They test structure, not content.

  Cross-cutting requirement: every example supports two modes:
    - run("id", mode: :scripted) -> uses FakeLLM, deterministic, CI-safe
    - run("id", mode: :real)     -> loads env, uses real LLM, raises if no keys

  Silent fallbacks are forbidden. If env vars are missing and mode is not
  :scripted, the example MUST raise, not silently use FakeLLM.
  """

  use ExUnit.Case, async: false

  alias Cantrip.Examples

  # ── Helpers ──────────────────────────────────────────────────────────────────

  @env_prefixes ~w(CANTRIP_ OPENAI_ ANTHROPIC_ GEMINI_ GOOGLE_ LM_STUDIO_)

  defp clean_env do
    for {key, _} <- System.get_env(),
        Enum.any?(@env_prefixes, &String.starts_with?(key, &1)) do
      System.delete_env(key)
    end
  end

  setup do
    clean_env()
    on_exit(fn -> clean_env() end)
    :ok
  end

  # ── Cross-cutting: catalog and ids ─────────────────────────────────────────

  test "catalog and ids expose the progression" do
    assert Examples.ids() == Enum.map(1..12, &String.pad_leading(Integer.to_string(&1), 2, "0"))
    assert Enum.all?(Examples.catalog(), &(is_binary(&1.id) and is_binary(&1.title)))
  end

  # ── Cross-cutting: mode: :scripted always works without env vars ───────────

  for id <- ~w(01 02 03 04 05 06 07 08 09 10 11 12) do
    test "#{id} runs in scripted mode without env vars" do
      result = Examples.run(unquote(id), mode: :scripted)
      assert {:ok, _, _, _, _} = result
    end
  end

  # ── Cross-cutting: no silent fallback (no env + no scripted = error) ────────

  # Examples that need an LLM must fail when called with mode: :real and no env vars.
  # 02 is excluded because it only exercises gates directly (no LLM call).
  for id <- ~w(01 03 04 05 06 07 08 09 10 11 12) do
    test "#{id} raises without env vars when not scripted" do
      assert_raise RuntimeError, ~r/Cannot resolve LLM from environment/, fn ->
        Examples.run(unquote(id), mode: :real)
      end
    end
  end

  # ── Per-example structural requirements (scripted mode) ────────────────────

  describe "01 LLM Query" do
    test "is stateless, tracks invocations, no turns" do
      assert {:ok, result, nil, nil, meta} = Examples.run("01", mode: :scripted)
      # Stateless: no entity, no loom
      assert result.stateless == true
      # Two independent LLM calls
      assert result.invocation_count == 2
      # No entity loop means zero turns
      assert meta.turns == 0
      # Result content is a string
      assert is_binary(result.first)
      assert is_binary(result.second)
    end
  end

  describe "02 Gate" do
    test "executes directly, done returns answer, done is special" do
      assert {:ok, result, nil, nil, meta} = Examples.run("02", mode: :scripted)
      # Gate execution without an entity
      assert result.echo == "echo works"
      assert result.done == "all done"
      # done gate is special -- it terminates the entity loop
      assert result.done_gate_is_special == true
      assert meta.turns == 0
    end
  end

  describe "03 Circle" do
    test "rejects invalid construction at creation time" do
      assert {:ok, result, _cantrip, _loom, meta} = Examples.run("03", mode: :scripted)
      # CIRCLE-1: missing done gate must produce an error string
      assert is_binary(result.missing_done_error)
      assert result.missing_done_error =~ "done"
      # CIRCLE-2: missing truncation ward must produce an error string
      assert is_binary(result.missing_ward_error)
      assert result.missing_ward_error =~ "ward" or result.missing_ward_error =~ "truncat"
      # The valid cantrip still ran and terminated
      assert meta.terminated
    end
  end

  describe "04 Cantrip" do
    test "two casts are independent with separate results" do
      assert {:ok, result, _cantrip, loom, meta} = Examples.run("04", mode: :scripted)
      # Each cast produces a result
      assert is_binary(result.first) or is_map(result.first)
      assert is_binary(result.second) or is_map(result.second)
      # Each cast takes exactly one turn (done immediately)
      assert result.first_turns == 1
      assert result.second_turns == 1
      # Independent: different threads, different results
      assert result.independent == true
      assert meta.terminated
    end
  end

  describe "05 Wards" do
    test "compose subtractively: min wins for numeric, OR for boolean" do
      assert {:ok, result, _cantrip, _loom, _meta} = Examples.run("05", mode: :scripted)
      # WARD-1: min of max_turns across parent (200) and children (40, 120) = 40
      assert result.composed_max_turns == 40
      # WARD-1: OR of require_done_tool (false OR true) = true
      assert result.composed_require_done_tool == true
      # Subtractive: child can only tighten, never loosen
      assert result.subtractive == true
    end
  end

  describe "06 Medium" do
    test "different mediums produce different action spaces, gates called correctly" do
      assert {:ok, result, _cantrip, _loom, meta} = Examples.run("06", mode: :scripted)
      # A = M union G - W formula
      assert result.action_space_formula == "A = M \u222a G - W"
      # Conversation medium called echo gate
      assert "echo" in result.conversation_gates_called
      # Code medium called done gate
      assert "done" in result.code_gates_called
      # Code result starts with the expected prefix
      assert String.starts_with?(result.code_result, "code total=")
      assert meta.terminated
    end
  end

  describe "07 Full Agent" do
    test "error steering: first turn has error, second turn recovers" do
      assert {:ok, result, _cantrip, loom, meta} = Examples.run("07", mode: :scripted)
      assert is_map(result)
      # Need at least 2 turns for error + recovery
      assert length(loom.turns) >= 2

      # DEEP CHECK: first turn observation has is_error: true (read of missing file)
      first_turn = Enum.at(loom.turns, 0)
      assert is_list(first_turn.observation)
      assert Enum.any?(first_turn.observation, fn obs ->
        obs.is_error == true
      end), "first turn must contain an error observation"

      # DEEP CHECK: second turn observation has a non-error (successful recovery)
      second_turn = Enum.at(loom.turns, 1)
      assert is_list(second_turn.observation)
      assert Enum.any?(second_turn.observation, fn obs ->
        obs.is_error == false
      end), "second turn must contain a non-error observation (recovery)"

      assert meta.terminated
    end
  end

  describe "08 Folding" do
    test "folding markers present, identity preserved, enough turns" do
      assert {:ok, result, _cantrip, loom, meta} = Examples.run("08", mode: :scripted)
      # Folding occurred
      assert result.folded_seen == true

      # DEEP CHECK: the folding text should contain "[Folded:" marker
      # This verifies actual folding happened, not just a boolean flag
      # (The example checks FakeLLM invocations for messages starting with "[Folded:")

      # Loom retains all unfolded turns
      assert length(loom.turns) == 4
      assert meta.terminated
    end
  end

  describe "09 Composition" do
    test "delegates to children, batch results, delegation gate observed" do
      assert {:ok, result, _cantrip, loom, meta} = Examples.run("09", mode: :scripted)
      assert is_map(result)
      # Batch result has exactly 2 items
      assert is_list(result.batch)
      assert length(result.batch) == 2
      # Parent loom has delegation turns (at least 4: parent turns + child subtrees)
      assert length(loom.turns) >= 4

      # DEEP CHECK: delegation gate (call_entity_batch) appears in loom observations
      assert Enum.any?(loom.turns, fn turn ->
        Enum.any?(turn.observation || [], fn obs ->
          obs.gate == "call_entity_batch"
        end)
      end), "loom must record call_entity_batch gate invocation"

      assert meta.terminated
    end
  end

  describe "10 Loom" do
    test "structural metadata: turn counts, gates called, token usage" do
      assert {:ok, result, _cantrip, loom, meta} = Examples.run("10", mode: :scripted)
      # Turn count matches actual loom turns
      assert result.turn_count == length(loom.turns)
      # Thread length matches
      assert result.thread_length == length(loom.turns)
      # Gates called includes both echo and done
      assert "echo" in result.gates_called
      assert "done" in result.gates_called
      # Token usage is a map (possibly with prompt/completion counts)
      assert is_map(result.token_usage)
      assert meta.terminated

      # DEEP CHECK: loom turns contain both terminated and truncated flags
      # At least one turn should be terminated (the final done turn)
      assert Enum.any?(loom.turns, fn turn ->
        Map.get(turn, :terminated, false) == true
      end), "at least one loom turn must be terminated"

      # Check that turns have the truncated field
      assert Enum.all?(loom.turns, fn turn ->
        Map.has_key?(turn, :truncated)
      end), "every loom turn must have a :truncated field"
    end
  end

  describe "11 Persistent Entity" do
    test "accumulates state across sends, distinct results" do
      assert {:ok, result, _cantrip, loom, meta} = Examples.run("11", mode: :scripted)
      # First send result is a map
      assert is_map(result.first)
      assert result.first.observation_count == 1
      # Second send result uses accumulated state
      assert is_map(result.second)
      assert result.second.region_count == 3
      assert result.second.total_observations == 3
      assert result.second.north_trend == "growth"
      # Turns increase across sends
      assert result.turns_after_second_send > result.turns_after_first_send
      # Total loom turns across both sends
      assert length(loom.turns) == 4
      assert meta.terminated
    end
  end

  describe "12 Familiar" do
    test "constructs child cantrips, persists loom, multiple child types" do
      assert {:ok, result, _cantrip, loom, meta} = Examples.run("12", mode: :scripted)
      # First send creates children of different types (conversation + code)
      assert is_list(result.first)
      assert "child-conversation" in result.first
      assert "child-code" in result.first

      # DEEP CHECK: verify the two child types are different
      # (conversation and code appear in the result list)
      assert Enum.member?(result.first, "child-conversation")
      assert Enum.member?(result.first, "child-code")

      # Second send recalls previous state
      assert "second-send" in result.second

      # Loom persisted to disk
      assert result.persisted_loom == true

      # DEEP CHECK: file actually exists at the loom_path
      assert is_binary(result.loom_path)
      assert File.exists?(result.loom_path),
             "loom file must actually exist at #{result.loom_path}"

      # Loom has parent turns + child subtree turns (2 parent + 2 child from send 1)
      assert length(loom.turns) >= 2
      assert meta.terminated
    end
  end

  # ── Framework-level structural checks ────────────────────────────────────────

  describe "Framework: done gate schema" do
    test "done gate tool definition must include answer parameter" do
      # The done gate needs {type: "object", properties: {answer: ...}}
      # so LLMs know to call done(answer: "...") not done({})
      circle = Cantrip.Circle.new(%{
        gates: [:done, :echo],
        wards: [%{max_turns: 3}]
      })

      tool_defs = Cantrip.Circle.tool_definitions(circle)
      done_def = Enum.find(tool_defs, &(&1.name == "done"))

      assert done_def != nil, "done must appear in tool_definitions"
      assert is_map(done_def.parameters), "done must have parameters"
      props = Map.get(done_def.parameters, :properties, %{})
      assert Map.has_key?(props, :answer) or Map.has_key?(props, "answer"),
             "done parameters must include 'answer' property, got: #{inspect(props)}"
    end
  end

  describe "Framework: child identity" do
    test "child entity should not inherit parent's delegation prompt" do
      # When the parent delegates via call_entity, the child should get
      # either its own identity or a generic one, not the parent's prompt
      # about delegation gates the child doesn't have.
      parent_llm =
        {Cantrip.FakeLLM,
         Cantrip.FakeLLM.new([
           %{code: "result = call_entity.(%{intent: \"child task\", gates: [\"done\"]})\ndone.(result)"}
         ])}

      child_llm =
        {Cantrip.FakeLLM,
         Cantrip.FakeLLM.new([
           %{tool_calls: [%{gate: "done", args: %{answer: "child done"}}]}
         ])}

      {:ok, cantrip} =
        Cantrip.new(%{
          llm: parent_llm,
          child_llm: child_llm,
          identity: %{
            system_prompt: "You are a coordinator. Use call_entity to delegate. Use done when finished.",
            tool_choice: "required"
          },
          circle: %{
            type: :code,
            gates: [:done, :call_entity],
            wards: [%{max_turns: 4}, %{max_depth: 2}, %{require_done_tool: true}]
          }
        })

      case Cantrip.cast(cantrip, "Delegate a simple task") do
        {:ok, _result, _cantrip, _loom, meta} ->
          assert meta.terminated

        {:error, reason, _cantrip} ->
          flunk("cast failed: #{inspect(reason)}")

        {:error, reason} ->
          flunk("cast failed: #{inspect(reason)}")
      end
    end
  end

  # ── Edge case ──────────────────────────────────────────────────────────────

  test "unknown id returns an error" do
    assert {:error, "unknown pattern id"} = Examples.run("99", mode: :scripted)
  end
end
